R/estimate_contrasts.R

```diff
@@ -135,6 +135,31 @@ estimate_contrasts <- function(model,
   contrasts$contrast <- NULL
   contrasts <- cbind(level_cols, contrasts)
 
+  # Add effect size (Cohen's d)
+  if (is.null(contrast) && is.null(fixed) && is.null(at)) {
+    dat <- insight::get_data(model)
+    resp <- insight::find_response(model)
+
+    if (is.numeric(dat[[resp]])) {
+      dat <- datawizard::data_select(dat, c(resp, info$contrast))
+
+      list.dat <- lapply(seq_len(nrow(contrasts)), function(i) {
+        log.vec <- which(dat[[info$contrast]] == unlist(info$misc$orig.grid)[[i]])
+        dat.temp <- datawizard::data_filter(dat, log.vec)
+      })
+      list.dat <- stats::setNames(list.dat, unique(unlist(level_cols)))
+
+      eff <- lapply(seq_len(nrow(contrasts)), function(i) {
+        effectsize::cohens_d(x = list.dat[[contrasts$Level1[i]]][[resp]],
```

Comment:
I wonder if getting raw Cohen's *d* like that from the raw data instead of relying on the marginal statistics isn't misleading though 🤔 @mattansb @bwiernik what do you think?

Reply:
Neither option is ideal (see discussion here: https://github.com/easystats/effectsize/issues/212, and note that `cohens_d()` has a `paired = ` argument).

I think you're looking for this https://github.com/easystats/effectsize/issues/351?

Reply:
My approach to calculating effect sizes in [`rempsyc::nice_contrasts`](https://rempsyc.remi-theriault.com/reference/nice_contrasts.html) is to use [`bootES::bootES`](https://github.com/dgerlanc/bootES/blob/main/R/bootES.R), which allows specifying contrasts (and a robust version of Cohen's _d_, and other cool things).

For `modelbased`, I wanted to stay within the easystats ecosystem to avoid adding another dependency, so I went for the regular Cohen's _d_ for the short-term, forgetting about easystats/effectsize#351. But if I understand correctly, #351 is not ready.

So that leaves us four choices for the _short-term_:

1. Stick with Cohen's _d_, even if it is not ideal
2. Derive it from the marginal statistics, like Dom suggested
3. Use `bootES::bootES` like in `rempsyc`
4. Merge #351 so I can use that.

Obviously no 4 is best, but I don't know what is missing so we can merge it? If we have to wait for ever for it to be fixed (it's been 2 years already), I'd rather have a short-term solution. So what do you want me to do short-term @mattansb @DominiqueMakowski ?

Reply:
We can compute legit cohens d from the coefficients/marginal contrasts by adjusting the sigma to add back any other covariate variance that has been pulled out

Reply:
For anything other than a linear regression with a single factor predictor, computing a standardized difference requires decisions (do we adjust for the covariances or not? Do we standardize by the residual sd or do we also account for level-2 variance(s)?).
There are not trivial decisions, so I suggest *not* including these by default (or at all).

Maybe let's be part of the movement that encourages people to think about their X/Y scales? The movement to mean people off of standardized effect sizes?

Reply:
(saw this, and will reply once the situation here gets back to normal ✌️)

Reply:
> Maybe let's be part of the movement that encourages people to think about their X/Y scales? The movement to mean people off of standardized effect sizes?

I’m all for using original scale units when those are meaningful. But in my department for example, almost everyone relies exclusively on Likert scales, many questionnaire also use a different number of response options. In this context, do you still think it makes more sense to use non-standardized scores?

> (do we adjust for the covariances or not? Do we standardize by the residual sd or do we also account for level-2 variance(s)?) There are not trivial decisions, so I suggest not including these by default (or at all).

I’d have to agree with Dom here. A lot of people argue that contrast analysis is superior to plain _t_ tests. But with _t_ tests at least you can get Cohen’s _d_. So it seems like there is no point to do contrasts analysis if there is not yet a satisfying solution to get the effect size.

So what will people do? Take the next best option. My colleagues are likely to simply attempt to compute Cohen’s _d_ on the groups directly, even if this is mistaken with contrasts. Slightly more savvy people like me might think of using `bootES::bootES()` or `emmeans::eff_size()`. I guess that’s the best currently implemented options.

So I think providing a reasonable default with some decisions taken for the users, yes, it will actually prevent greater harm (people doing Cohen's _d_ by hand because no effect size is provided).

I think it's not always reasonable to expect most people to understand stats at such a """deep""" level e.g., for that kind of decision-making. These are not trivial decisions, granted. But I'm pretty sure nobody at my department would even know what decision to take there. They wouldn't even know what it means or the implications. I mean I can't blame them. They have to become experts as therapists and clinicians, at the theories of psychology, at methodology, at statistical analysis, at writing, at public speaking, at managing others, at marketing themselves, etc., AND they might have to work outside school, or they might be volunteering in student clubs or sports or whatnot. That's a lot to ask :)

So if we can't even agree on the effect size here, imagine those regular non-savvy students! Other fields frequently have a dedicated statistician to their research team. But (sadly?), in psychology, this is not the norm. So I think that's why people like `easystats` so much. It makes it easy to get "good enough" outputs by default. So as long as our decision-making is documented and can be overwritten, I think we're OK.

> Is whatever bootES() provides legit?

Well, after double-checking, even though it supports contrasts, it works with the original data frame and the contrasts provided, not with the `lm` model, so I think it would not take into account all the other stuff specified in the model like covariates, etc.

> We can compute legit cohens d from the coefficients/marginal contrasts by adjusting the sigma to add back any other covariate variance that has been pulled out

Ultimately, I think Brenton’s solution would be the most realistic: best compromise of effort and output acceptability. I just don’t understand what it means and how I could implement it myself sadly 😬

Reply:
> I hear that, but I still think adding an option to get reasonable effect size (by default) would be good.

If you review the lit, I think you will find that people almost never report Cohen's d for contrasts, and when they do it is not clear what standardization method they are using, so there is no "reasonable" option that is required or available. If ***we*** decide to give some measure of effect size, we will be setting a *new* default.

> Now as to what a "reasonable" ES is in that case, well I would naively say "whatever is the most natural extension of a traditional Cohen's d".

Here, then, I think I disagree with @bwiernik - the natural extension would use the residual sigma for the standardization, giving a standardized difference between conditional population means.

> I don't know what the answer there, but returning to the initial purpose of effectsize it was mostly to have something to give an intuitive "feel" of effect magnitude. [...]

I completely disagree with this whole paragraph 😅. If you want to give people an intuitive feel for the size of their contrasts, it is SUPER IMPORTANT that you compute a proper effect size! If we want to give people what they *want* from contrasts, I'd argue they only want the sign (positive of negative difference) and a p-value.... If ***we*** decide to give some measure of effect size, we are responsible for giving *a good one*.

---

> I’m all for using original scale units when those are meaningful. But in my department for example, almost everyone relies exclusively on Likert scales, many questionnaire also use a different number of response options. In this context, do you still think it makes more sense to use non-standardized scores?

Absolutely! Within the context of a 1-7 scale, a change of 1.2 units is much more understandable to me than "a change of 0.4 pooled SDs".

> I’d have to agree with Dom here. A lot of people argue that contrast analysis is superior to plain t tests. But with t tests at least you can get Cohen’s d. So it seems like there is no point to do contrasts analysis if there is not yet a satisfying solution to get the effect size.

You can say the same thing about regression - regression is superior to a correlation, yet the standardized coefficients are harder to interpret (can be larger than 1, for example), so that makes the method less useful? No, it just means that simple problems have simple solutions.

But here we can agree - if you are doing a one-way (between subjects) ANOVA instead of 16 t-tests, it is trivial to compute Cohen's d for the contrasts - the estimated difference divided by the pooled sd.

... I personally prefer people do Cohen's d by hand. I see no reason not to do that. If your design is simple enough, the only difference between a t-test and contrasts is the SE, the raw estimate will be the same.

---

The ideal solution to me would be:
1. No standardized effect sizes by default.
2. Have **several** methods for standardized effect sizes (essentially wrappers around something like `emmeans::eff_size()` to make use of it easy) for how one can standardize their estimates + a vignette explaining the differences.

This can cover a range of methods, such as https://github.com/easystats/effectsize/issues/351 or even  https://github.com/easystats/effectsize/issues/372 for rm data / mixed models.

This would be an effort I can get behind.

Reply:
I think that's reasonable and I can get behind your suggestion too. So then let's not rush things just to have something short-term (so I'll close this PR since that was the original objective). We can probably start working on implementing these solutions bit by bit.

> No standardized effect sizes by default.

Would that mean _no_ effect sizes by default or _non-standardized_ effect sizes by default? And what does this look like? I suppose it's not just the mean difference since like you were saying we may need to adjust for other stuff in the model like covariates and other predictors? So just adjusted mean difference?

Reply:
> - No standardized effect sizes by default.
> Have several methods for standardized effect sizes (essentially wrappers around something like emmeans::eff_size() to make use of it easy) for how one can standardize their estimates + a vignette explaining the differences.

That sounds fair

Reply:
> I think this is just the Difference column in the current output haha ("marginal means"), that's the (non-standardized) effect size.

Yup.

YAY!